### PR TITLE
[drape] SetCenterEvent animation fix

### PR DESCRIPTION
**Pull request by AndrewShkrob**
Android Auto before

https://user-images.githubusercontent.com/10351358/215907976-88d3f31e-2bc3-42d7-a487-9621db6f1b74.mov

Android Auto after

https://user-images.githubusercontent.com/10351358/215907991-88cf49b7-af39-48f9-b1b3-2941e37faa0b.mov

Android before

https://user-images.githubusercontent.com/10351358/215908077-7fead4f0-13e2-4407-b2df-2712d0fe3e52.mp4

Android after

https://user-images.githubusercontent.com/10351358/215908098-52cfbfdf-29f8-427a-be67-ff3b3eae1d39.mp4

### CODE REVIEW

**Comment by biodranik** [comment]
Cool, thanks!

1. How it will look without animation?
2. How it will look with 2x faster animation?

**Comment by AndrewShkrob** [comment]
@biodranik 
> How it will look without animation?

https://user-images.githubusercontent.com/10351358/216082268-34cdde59-dfb4-4342-a3e5-9820da02b2ef.mp4

> How it will look with 2x faster animation?

https://user-images.githubusercontent.com/10351358/216082385-0f39006c-f5d0-41c3-aebc-d6caf874cfec.mp4

File: drape_frontend/user_event_stream.cpp [hunk 1]
```diff
@@ -44,6 +44,49 @@ size_t GetValidTouchesCount(std::array<Touch, 2> const & touches)
...
+
+  anim->SetOnFinishAction(onFinishAction);
+
+  if (parallelAnimCreator != nullptr)
+  {
+    drape_ptr<ParallelAnimation> parallelAnim = make_unique_dp<ParallelAnimation>();
+    parallelAnim->SetCustomType(kParallelFollowAnim);
+    parallelAnim->AddAnimation(parallelAnimCreator(anim->GetType() == Animation::Type::MapFollow ? make_ref(anim)
+                                                                                                 : nullptr));
+    parallelAnim->AddAnimation(move(anim));
```
Comment (biodranik) [inline drape_frontend/user_event_stream.cpp:82]:
```suggestion
    parallelAnim->AddAnimation(std::move(anim));
```

File: drape_frontend/user_event_stream.cpp [hunk 2]
```diff
@@ -44,6 +44,49 @@ size_t GetValidTouchesCount(std::array<Touch, 2> const & touches)
...
+    return parallelAnim;
+  }
+  else
+  {
+    return anim;
+  }
```
Comment (biodranik) [inline drape_frontend/user_event_stream.cpp:88]:
```suggestion
  return anim;
```

File: drape_frontend/user_event_stream.cpp [hunk 3]
```diff
@@ -409,6 +452,15 @@ bool UserEventStream::OnSetCenter(ref_ptr<SetCenterEvent> centerEvent)
   {
     screen.SetFromParams(center, screen.GetAngle(), GetScreenScale(zoom));
     screen.MatchGandP3d(center, m_visibleViewport.Center());
+    if (centerEvent->IsAnim())
+    {
+      ShrinkAndScaleInto(screen, df::GetWorldRect());
+      drape_ptr<Animation> anim = createMapFollowAnimation(
+        GetCurrentScreen(), screen, center, m_visibleViewport.Center(),
+        zoom,screen.GetAngle(), false);
```
Comment (biodranik) [inline drape_frontend/user_event_stream.cpp:460]:
```suggestion
      drape_ptr<Animation> anim = createMapFollowAnimation(
          GetCurrentScreen(), screen, center, m_visibleViewport.Center(),
          zoom, screen.GetAngle(), false);
```

File: drape_frontend/user_event_stream.cpp [hunk 4]
```diff
@@ -600,42 +652,10 @@ bool UserEventStream::SetFollowAndRotate(m2::PointD const & userPos, m2::PointD
...
-      parallelAnim->AddAnimation(move(anim));
-      m_animationSystem.CombineAnimation(move(parallelAnim));
-    }
-    else
-    {
-      m_animationSystem.CombineAnimation(move(anim));
-    }
+    drape_ptr<Animation> anim = createMapFollowAnimation(
+      currentScreen, screen, userPos, pixelPos, preferredZoomLevel,
+      -azimuth, isAutoScale, onFinishAction, parallelAnimCreator);
```
Comment (biodranik) [inline drape_frontend/user_event_stream.cpp:657]:
```suggestion
    drape_ptr<Animation> anim = createMapFollowAnimation(
        currentScreen, screen, userPos, pixelPos, preferredZoomLevel,
        -azimuth, isAutoScale, onFinishAction, parallelAnimCreator);
```

Comment (biodranik) [reply drape_frontend/user_event_stream.cpp:657]: Two spaces mean a different operation. 4 spaces — the same multiline operation.

**Review by biodranik — COMMENTED** [review]
2x animation looks awesome! Will it be 2x everywhere in the app? Or only in this scenario? What other animations are there? Let's make them all 2x faster!

@vng PTAL

**Comment by AndrewShkrob** [comment]
> 2x animation looks awesome! Will it be 2x everywhere in the app? Or only in this scenario? What other animations are there? Let's make them all 2x faster!

Currently, it is used in 2 cases:
1. Switching to FOLLOW mode with scale
2. Switching to FOLLOW_AND_ROTATE mode

We can speed up animation only for the 1st case, but it also looks good for the 2nd.
For the switching to FOLLOW mode, there are 2 other cases:

Only scale - it's never used. Possibly, it can be removed from the code
https://github.com/AndrewShkrob/organicmaps/blob/c9694e5b5a7b78537a0bc2eec44dd9658b8e7a07/drape_frontend/user_event_stream.cpp#L465-L469

2. Only move - used when we are on the same scale

https://user-images.githubusercontent.com/10351358/216094478-e3093c02-ca5b-4f88-877c-3d88f8f48a72.mp4

I can also try to speed up this one.

**Comment by AndrewShkrob** [comment]
> What about the CarPlay/AA case, or when users press + / - zoom buttons? Shouldn't it be used there?

It's the other event - ScaleEvent.

> Regarding the last one, can it be converted to move-only, to look and work faster? That zooming-out and zooming-in is really annoying.

I'll try.

**Comment by endim8** [comment]
Also regarding the last one, there's an issue open for it: https://github.com/organicmaps/organicmaps/issues/1235

the current movement is: (Move Up), (move XY), (Move Down)
the suggested movement is:  (Move up & half XY), (move down and half XY)

**Review by vng — APPROVED** [review]

**Comment by AndrewShkrob** [comment]
> I don't think the current implementation can do both animations at the same time. Right?

It's possible. Unfortunately, I didn't have any luck. It's hard to understand transformations logic.
Here is how it may look like without scale animation:

https://user-images.githubusercontent.com/10351358/216774221-a10f24e4-cc33-4fc1-abd9-dfcf53f64381.mp4


https://user-images.githubusercontent.com/10351358/216774223-4069001c-6e56-49d9-9487-aa635d18f1cf.mp4


I'll update the PR if it's OK

**Comment by biodranik** [comment]
Argh, it is worse than without animation at all.

**Comment by AndrewShkrob** [comment]
Ok. Let's leave it as it is but I need to check one more thing before the merge

**Comment by AndrewShkrob** [comment]
@biodranik I won't change anything else now. Let's merge it

**Comment by biodranik** [comment]
Reverted this change because of a bug in the animation when a search result is selected.

**Pull request merged**